### PR TITLE
Starting infrastructure to simulate and build bams for more rigorous summarizer testing

### PR DESCRIPTION
**Pull request by drakeeee**
TODO:
- ~~Add sam flags specifications~~
- ~~Ensure proper behavior of reverse read~~

The goal with this is to begin making specific functions that assess corner cases and the behavior of methods involved in bam -> bed -> counting.

### CODE REVIEW

File: fbio/bam/build_bam_utils.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        template_length (int):
+        reference_fasta (str): path to reference fasta if using a test fasta
+    """
+    def __init__(self, chrom, start, stop, query_name=None,
+                 n_soft_clips=True, is_reverse=None, is_proper_pair=None,
+                 is_paired=None, reference_id=None, mate_reference_id=None,
+                 mapping_quality=0, cigarstring=None, is_read1=None,
+                 query_qualities=None, start_soft_clip_length=0, stop_soft_clip_length=0,
+                 mate_start=None, template_length=0, reference_fasta=None):
+        assert isinstance(is_read1, bool), "read1 bool must be set"
```
Comment (brandon-white) [inline fbio/bam/build_bam_utils.py]: Why are these not positional args if they must be set?

File: fbio/bam/build_bam_utils.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        assert isinstance(is_reverse, bool), "is_reverse must be set"
+        assert isinstance(is_proper_pair, bool), "is_proper_pair must be set"
+
+        if is_proper_pair:
+            assert isinstance(is_paired, bool), "is_paired must be set"
+
+        logging.warning('{} {} {}'.format(chrom, start, stop))
```
Comment (brandon-white) [inline fbio/bam/build_bam_utils.py]: should this be a debug or removed?

File: fbio/bam/build_bam_utils.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        length = (stop - start)
+
+        read = pysam.AlignedSegment()
+        read.reference_start = start
+        read.query_name = query_name
+        read.mapping_quality = mapping_quality
+
+        # awaiting a pysam update
+        # read.reference_name = chrom
+        read.reference_id = reference_id
```
Comment (brandon-white) [inline fbio/bam/build_bam_utils.py]: Can you add a # TODO?

File: fbio/bam/build_bam_utils.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+            read.is_read2 = True
+
+        read.is_reverse = is_reverse
+        read.is_proper_pair = is_proper_pair
+        if is_proper_pair:
+            read.is_paired = is_paired
+            read.mate_is_reverse = not is_reverse
+
+        if is_paired:
+            # awaiting a pysam update
```
Comment (brandon-white) [inline fbio/bam/build_bam_utils.py]: Can you add a # TODO or remove?

File: fbio/bam/build_bam_utils.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        str: sequence string
+    """
+    if reference_fasta is not None:
+        fasta = pysam.FastaFile(reference_fasta)
+    else:
+        fasta = pysam.FastaFile(REFERENCE_FASTA)
+
+    sequence = fasta.fetch(chrom, start, stop)
+
+    logging.warning('seq: {}'.format(sequence))
```
Comment (brandon-white) [inline fbio/bam/build_bam_utils.py]: debug or warning?

File: fbio/util/gtf_utils.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+from fbio.region import Region
```
Comment (brandon-white) [inline fbio/util/gtf_utils.py]: Remove this, its not used? We also have one in summarizer.

File: fbio/test/bam/test_build_bam_utils.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import numpy as np
+import pysam
+
+from fbio.bam.build_bam_utils import Fragment
+
+
+TEST_DATA_DIR = path.join(path.dirname(path.abspath(__file__)), 'data')
+TEST_BAM_PATH = '{}/align.bam'.format(TEST_DATA_DIR)
+TEST_FASTA = '{}/test.fasta'.format(TEST_DATA_DIR)
+
```
Comment (brandon-white) [inline fbio/test/bam/test_build_bam_utils.py]: Can you add a test to ensure the simulator produces like 50-100 correct fragments?

File: fbio/bam/build_bam_utils.py (hunk 1 above)
Comment (drakeeee) [reply fbio/bam/build_bam_utils.py]: good question

File: fbio/bam/build_bam_utils.py (hunk 2 above)
Comment (drakeeee) [reply fbio/bam/build_bam_utils.py]: debug

File: fbio/util/gtf_utils.py (hunk 6 above)
Comment (drakeeee) [reply fbio/util/gtf_utils.py]: the difference between this one and the one in summarizer is that summarizer's loads GTFs while this one makes them. I'll demonstrate use cases.

**Pull request merged**